File: apps/api-go/internal/handler/ai.go
package handler

import (
	"fmt"
	"time"

	"github.com/alldata/api-go/pkg/response"
	"github.com/gin-gonic/gin"
)

// AIChat AI 对话
func AIChat(c *gin.Context) {
	var req struct {
		Messages []struct {
			Role    string `json:"role"`
			Content string `json:"content"`
		} `json:"messages" binding:"required"`
		Model  string `json:"model"`
		Stream bool   `json:"stream"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "参数错误: "+err.Error())
		return
	}

	// 模拟 AI 响应
	result := map[string]interface{}{
		"content": fmt.Sprintf("这是 AI 助手对您问题的回复。您发送了 %d 条消息。", len(req.Messages)),
		"model":   req.Model,
		"usage": map[string]int{
			"prompt_tokens":     100,
			"completion_tokens": 50,
			"total_tokens":      150,
		},
		"created_at": time.Now().Format(time.RFC3339),
	}
	response.Success(c, result)
}

// AIComplete AI 文本完成
func AIComplete(c *gin.Context) {
	var req struct {
		Messages []struct {
			Role    string `json:"role"`
			Content string `json:"content"`
		} `json:"messages" binding:"required"`
		Model string `json:"model"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "参数错误: "+err.Error())
		return
	}

	result := map[string]interface{}{
		"content": "这是 AI 完成的文本内容。",
		"model":   req.Model,
		"usage": map[string]int{
			"prompt_tokens":     80,
			"completion_tokens": 30,
			"total_tokens":      110,
		},
	}
	response.Success(c, result)
}

// AIModels 获取可用模型列表
func AIModels(c *gin.Context) {
	models := []map[string]interface{}{
		{"id": "gpt-4", "name": "GPT-4", "provider": "openai", "max_tokens": 8192},
		{"id": "gpt-3.5-turbo", "name": "GPT-3.5 Turbo", "provider": "openai", "max_tokens": 4096},
		{"id": "claude-3", "name": "Claude 3", "provider": "anthropic", "max_tokens": 100000},
	}
	response.Success(c, models)
}

// AIConfig 获取 AI 配置（脱敏）
func AIConfig(c *gin.Context) {
	config := map[string]interface{}{
		"provider":        "openai",
		"model":           "gpt-4",
		"max_tokens":      4096,
		"temperature":     0.7,
		"api_key_masked":  "sk-****...****",
		"enabled":         true,
	}
	response.Success(c, config)
}

// AIInsight 生成智能洞察
func AIInsight(c *gin.Context) {
	var req struct {
		ProjectID int    `json:"project_id" binding:"required"`
		DataType  string `json:"data_type" binding:"required"`
		DataID    int    `json:"data_id" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "参数错误: "+err.Error())
		return
	}

	result := map[string]interface{}{
		"summary":       "数据整体呈上升趋势，核心指标表现良好。",
		"trend":         "upward",
		"confidence":    0.85,
		"key_findings":  []string{"DAU 持续增长 15%", "转化率提升 8%", "用户留存率稳定在 65%"},
		"recommendations": []string{"继续优化新用户引导流程", "关注高价值用户留存", "加强推送触达"},
		"generated_at":  time.Now().Format(time.RFC3339),
	}
	response.Success(c, result)
}

// AIAnomalyDetect 异常检测
func AIAnomalyDetect(c *gin.Context) {
	var req struct {
		ProjectID  int      `json:"project_id" binding:"required"`
		MetricName string   `json:"metric_name" binding:"required"`
		DataPoints []float64 `json:"data_points" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "参数错误: "+err.Error())
		return
	}
	if len(req.DataPoints) < 3 {
		response.BadRequest(c, "数据点数量不足，至少需要 3 个")
		return
	}

	// 简单的异常检测模拟
	anomalies := []map[string]interface{}{}
	mean := 0.0
	for _, v := range req.DataPoints {
		mean += v
	}
	mean /= float64(len(req.DataPoints))

	for i, v := range req.DataPoints {
		if v > mean*1.5 || v < mean*0.5 {
			anomalies = append(anomalies, map[string]interface{}{
				"index":     i,
				"value":     v,
				"expected":  mean,
				"deviation": (v - mean) / mean,
				"severity":  "high",
			})
		}
	}

	result := map[string]interface{}{
		"metric_name": req.MetricName,
		"anomalies":   anomalies,
		"total_points": len(req.DataPoints),
		"mean":        mean,
		"detected_at": time.Now().Format(time.RFC3339),
	}
	response.Success(c, result)
}

// AIAnomalyInterpret 异常解读
func AIAnomalyInterpret(c *gin.Context) {
	var req struct {
		Anomalies  []map[string]interface{} `json:"anomalies" binding:"required"`
		MetricName string                   `json:"metric_name" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "参数错误: "+err.Error())
		return
	}

	result := map[string]interface{}{
		"metric_name": req.MetricName,
		"interpretation": fmt.Sprintf("检测到 %d 个异常点，建议关注 %s 指标的变化趋势。", len(req.Anomalies), req.MetricName),
		"possible_causes": []string{
			"可能是由于系统升级导致的短暂波动",
			"可能与近期运营活动相关",
			"建议检查数据采集链路是否正常",
		},
		"suggestions": []string{
			"持续监控该指标 24 小时",
			"对比历史同期数据",
			"排查近期变更影响",
		},
		"interpreted_at": time.Now().Format(time.RFC3339),
	}
	response.Success(c, result)
}

File: apps/web/src/services-new/ai.ts
/**
 * AI 模块 API 服务
 * 封装 AI 对话、洞察、异常检测等接口
 */
import { get, post } from './request';
import { useAuthStore } from '@/stores/auth';
import { useGlobalStore } from '@/stores/global';

// ─── 类型定义 ─────────────────────────────

export interface ChatMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

export interface CompletionResponse {
  id: string;
  content: string;
  model: string;
  usage: { prompt_tokens: number; completion_tokens: number; total_tokens: number };
  finish_reason: string;
  created_at: string;
}

export interface ModelInfo {
  id: string;
  name: string;
  provider: string;
  max_tokens: number;
  description: string;
}

export interface InsightRequest {
  project_id: number;
  data_type: string;
  data_id: number;
  context?: string;
}

export interface InsightResponse {
  id: string;
  summary: string;
  key_findings: string[];
  recommendations: string[];
  confidence: number;
  created_at: string;
}

export interface KeyFinding {
  title: string;
  description: string;
  importance: 'high' | 'medium' | 'low';
  metric_value?: number;
}

export interface Recommendation {
  title: string;
  description: string;
  priority: 'high' | 'medium' | 'low';
  expected_impact?: string;
}

export interface AnomalyDetectionRequest {
  project_id: number;
  metric_name: string;
  data_points: Array<{ timestamp: string; value: number }>;
  sensitivity?: number;
}

export interface AnomalyPoint {
  timestamp: string;
  value: number;
  expected: number;
  severity: 'low' | 'medium' | 'high' | 'critical';
}

export interface AnomalyResult {
  index: number;
  timestamp: string;
  value: number;
  expected: number;
  z_score?: number;
  severity: 'low' | 'medium' | 'high' | 'critical';
  category: 'spike' | 'trend_shift' | 'seasonal_break' | 'gradual_drift';
  methods: string[];
}

export interface AnomalyDetectionResponse {
  id: string;
  anomalies: AnomalyResult[];
  total_anomalies: number;
  summary: string;
  created_at: string;
}

export interface AnomalyInterpretation {
  id: string;
  possible_causes: string[];
  recommendations: string[];
  severity_assessment: 'low' | 'medium' | 'high' | 'critical';
  detailed_analysis: string;
  created_at: string;
}

// ─── API 方法 ─────────────────────────────

export const aiApi = {
  /** AI 对话（非流式） */
  chat: (messages: ChatMessage[], options?: { model?: string; temperature?: number }) =>
    post<CompletionResponse>('/ai/chat', { messages, stream: false, ...options }),

  /** AI 文本完成 */
  complete: (prompt: string, options?: { model?: string; temperature?: number }) =>
    post<CompletionResponse>('/ai/complete', { prompt, ...options }),

  /** 获取可用模型列表 */
  getModels: () =>
    get<ModelInfo[]>('/ai/models'),

  /** 获取 AI 配置 */
  getConfig: () =>
    get<Record<string, unknown>>('/ai/config'),

  /** 生成完整智能洞察 */
  generateInsight: (data: InsightRequest) =>
    post<InsightResponse>('/ai/insight', data),

  /** 趋势摘要 */
  generateTrend: (data: Record<string, unknown>) =>
    post<{ summary: string }>('/ai/insight/trend', data),

  /** 关键发现 */
  generateFindings: (data: Record<string, unknown>) =>
    post<{ findings: KeyFinding[] }>('/ai/insight/findings', data),

  /** 行动建议 */
  generateRecommendations: (data: Record<string, unknown>) =>
    post<{ recommendations: Recommendation[] }>('/ai/insight/recommendations', data),

  /** 异常检测 */
  detectAnomalies: (data: AnomalyDetectionRequest) =>
    post<AnomalyDetectionResponse>('/ai/anomaly/detect', data),

  /** 异常解读 */
  interpretAnomalies: (data: { anomalies: AnomalyResult[]; metric_name: string; context?: string }) =>
    post<AnomalyInterpretation>('/ai/anomaly/interpret', data),

  /** 监控配置 */
  setupMonitor: (data: Record<string, unknown>) =>
    post('/ai/anomaly/monitor', data),
};

/**
 * SSE 流式对话
 * 使用 fetch API 实现流式读取
 */
export async function streamChat(
  messages: ChatMessage[],
  onChunk: (chunk: string) => void,
  options?: { model?: string; temperature?: number },
): Promise<void> {
  const authStore = useAuthStore.getState();
  const globalStore = useGlobalStore.getState();
  const baseUrl = import.meta.env.VITE_API_URL ?? '/api';

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  };
  if (authStore.token) {
    headers['Authorization'] = `Bearer ${authStore.token}`;
  }
  if (globalStore.currentProject) {
    headers['Project-Id'] = String(globalStore.currentProject.id);
  }

  const response = await fetch(`${baseUrl}/ai/chat`, {
    method: 'POST',
    headers,
    body: JSON.stringify({ messages, stream: true, ...options }),
  });

  if (!response.ok) {
    throw new Error(`AI 对话请求失败: ${response.status}`);
  }

  const reader = response.body?.getReader();
  if (!reader) throw new Error('无法读取流式响应');

  const decoder = new TextDecoder();
  let buffer = '';

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split('\n');
    buffer = lines.pop() || '';

    for (const line of lines) {
      if (line.startsWith('data: ')) {
        try {
          const data = JSON.parse(line.slice(6));
          if (data.content) {
            onChunk(data.content);
          }
        } catch {
          // 忽略解析错误
        }
      }
    }
  }
}
